libwacom/libwacom-database.c

```diff
@@ -279,8 +323,81 @@ parse_stylus_id(const char *str, WacomStylusId *id)
 	return ret;
 }
 
+static char *
+string_or_fallback (GKeyFile *keyfile, const char *group,
+		    const char *key, const char *fallback, GError *error)
+{
+	GError *local_error = NULL;
+	bool ret;
+
+	ret = g_key_file_has_key(keyfile, group, key, &local_error);
+	if (local_error) {
+		g_warning ("String fallback error: %s\n", local_error->message);
+		g_clear_error (&local_error);
+	}
+	if (ret)
+		return g_key_file_get_string(keyfile, group, key, &error);
+
+	return g_strdup(fallback);
+}
+
+static int
+int_or_fallback (GKeyFile *keyfile, const char *group,
+		 const char *key, int fallback, GError *error)
+{
+	GError *local_error = NULL;
+	bool ret;
+
+	ret = g_key_file_has_key(keyfile, group, key, &local_error);
+	if (local_error) {
+		g_warning ("int fallback error: %s\n", local_error->message);
+		g_clear_error (&local_error);
+	}
+
+	if (ret)
+		return g_key_file_get_integer(keyfile, group, key, &error);
+
+	return fallback;
+}
+
+static bool
+boolean_or_fallback (GKeyFile *keyfile, const char *group,
+		     const char *key, bool fallback, WacomStylus *stylus, GError *error)
+{
+	GError *local_error = NULL;
+	bool ret;
+
+	ret = g_key_file_has_key(keyfile, group, key, &local_error);
+	if (local_error) {
+		g_warning ("boolean fallback error: %s\n", local_error->message);
+		g_clear_error (&local_error);
+	}
+
+	if (ret)
+		return g_key_file_get_boolean(keyfile, group, key, &error);
+
+	return fallback;
+}
+
+static gchar **
+garray_to_string_list(GArray *array)
+{
+	WacomStylusId *id;
+
+	if (array == NULL)
+		return NULL;
+
+	GArray *a = g_array_new(true, false, array->len);
+	for (guint i = 0; i < array->len; i++) {
+		id = &g_array_index(array, WacomStylusId, i);
+		g_array_append_vals(a, g_strdup_printf("0x%08x", id->tool_id), 1);
+	}
+
+	return g_array_steal(a, (long unsigned int *)(&(a->len)));
```

Comment:
That cast can have interesting side-effects if `sizeof(a->len) != sizeof(long unsigned int)`, an easier way around this is:

```suggestion
	gsize len = a->len;
	return g_array_steal(a, len);
```

Reply:
That is giving me the following error. I'm not sure why g_array_steal wants a pointer. 
```
../libwacom/libwacom-database.c: In function ‘stylus_ids_as_hex’:
../libwacom/libwacom-database.c:397:33: error: passing argument 2 of ‘g_array_steal’ makes pointer from integer without a cast [-Wint-conversion]
  397 |         return g_array_steal(a, len);
      |                                 ^~~
      |                                 |
      |                                 gsize {aka long unsigned int}
In file included from /usr/include/glib-2.0/glib.h:33,
                 from ../libwacom/libwacomint.h:34,
                 from ../libwacom/libwacom-database.c:30:
/usr/include/glib-2.0/glib/garray.h:86:54: note: expected ‘gsize *’ {aka ‘long unsigned int *’} but argument is of type ‘gsize’ {aka ‘long unsigned int’}
   86 |                                    gsize            *len);
      |                                    ~~~~~~~~~~~~~~~~~~^~~
```

Reply:
Edited my suggestion above  - you can pass `&len` instead but after checking the source for `g_array_steal()` - len is an output-only argument and returns the number of elements in the array, we can pass NULL since we don't care about that value here.